DataFormats/Detectors/CTP/src/CTPRateFetcher.cxx

```diff
@@ -0,0 +1,153 @@
+{
+  auto triggerRate = fetchCTPratesClassesNoPuCorr(timeStamp, className, inputType);
+  if (triggerRate >= 0) {
+    return pileUpCorrection(triggerRate);
+  }
+  return -1;
+}
+double CTPRateFetcher::fetchCTPratesClassesNoPuCorr(uint64_t timeStamp, const std::string& className, int inputType)
+{
+  std::vector<ctp::CTPClass> ctpcls = mConfig->getCTPClasses();
+  std::vector<int> clslist = mConfig->getTriggerClassList();
+  int classIndex = -1;
+  for (size_t i = 0; i < clslist.size(); i++) {
+    if (ctpcls[i].name.find(className) != std::string::npos) {
+      classIndex = i;
+      break;
+    }
+  }
+  if (classIndex == -1) {
+    LOG(warn) << "Trigger class " << className << " not found in CTPConfiguration";
+    return -2.;
+  }
+  auto rate{mScalers->getRateGivenT(timeStamp * 1.e-3, classIndex, inputType)};
+  return rate.second;
+}
+double CTPRateFetcher::fetchCTPratesInputs(uint64_t timeStamp, int input)
+{
+  std::vector<ctp::CTPScalerRecordO2> recs = mScalers->getScalerRecordO2();
+  if (recs[0].scalersInps.size() == 48) {
+    return pileUpCorrection(mScalers->getRateGivenT(timeStamp * 1.e-3, input, 7).second);
+  } else {
+    LOG(error) << "Inputs not available";
+    return -1.;
+  }
+}
+double CTPRateFetcher::fetchCTPratesInputsNoPuCorr(uint64_t timeStamp, int input)
+{
+  std::vector<ctp::CTPScalerRecordO2> recs = mScalers->getScalerRecordO2();
+  if (recs[0].scalersInps.size() == 48) {
+    return mScalers->getRateGivenT(timeStamp * 1.e-3, input, 7).second;
+  } else {
+    LOG(error) << "Inputs not available";
+    return -1.;
+  }
+}
+
+double CTPRateFetcher::pileUpCorrection(double triggerRate)
+{
+  if (mLHCIFdata == nullptr) {
+    LOG(fatal) << "No filling" << std::endl;
+  }
+  auto bfilling = mLHCIFdata->getBunchFilling();
+  std::vector<int> bcs = bfilling.getFilledBCs();
+  double nbc = bcs.size();
+  double nTriggersPerFilledBC = triggerRate / nbc / constants::lhc::LHCRevFreq;
+  double mu = -std::log(1 - nTriggersPerFilledBC);
+  return mu * nbc * constants::lhc::LHCRevFreq;
+}
+
+void CTPRateFetcher::setupRun(int runNumber, o2::ccdb::BasicCCDBManager* ccdb, uint64_t timeStamp, bool initScalers)
+{
+  if (runNumber == mRunNumber) {
+    return;
+  }
+  mRunNumber = runNumber;
+  LOG(debug) << "Setting up CTP scalers for run " << mRunNumber;
+  std::map<string, string> metadata;
+  mLHCIFdata = ccdb->getSpecific<parameters::GRPLHCIFData>("GLO/Config/GRPLHCIF", timeStamp, metadata);
+  if (mLHCIFdata == nullptr) {
+    LOG(fatal) << "GRPLHCIFData not in database, timestamp:" << timeStamp;
+  }
+  metadata["runNumber"] = std::to_string(mRunNumber);
+  mConfig = ccdb->getSpecific<ctp::CTPConfiguration>("CTP/Config/Config", timeStamp, metadata);
+  if (mConfig == nullptr) {
+    LOG(fatal) << "CTPRunConfig not in database, timestamp:" << timeStamp;
+  }
+  if (initScalers) {
+    mScalers = ccdb->getSpecific<ctp::CTPRunScalers>("CTP/Calib/Scalers", timeStamp, metadata);
```

Comment:
Here you are fetching pointers on the objects in the ccdb cache, which still owns them. If the external ccdb object is used to fetch the same objects for different time stamp or initialized for different CCDB URL (like you do in the test), the old cache will be destroyed and these pointers will become invalid. You should probably use data member objects instead of pointers, i.e.

```
o2::parameters::GRPLHCIFData mLHCIFdata{}; // in the class declaration
//...
auto ptrHCIFdata = ccdb->getSpecific<parameters::GRPLHCIFData>("GLO/Config/GRPLHCIF", timeStamp, metadata);
if (ptrLHCIFdata == nullptr) {
  LOG(fatal) << "GRPLHCIFData not in database, timestamp:" << timeStamp;
}
mLHCIFdata = *ptrLHCIFdata;
```